lib/ocpp/v201/connectivity_manager.cpp

```diff
@@ -0,0 +1,748 @@
+            // TODO what to do here???
+        }
+    } else if (this->pending_network_slot != 0) {
+        // There is a network slot pending, maybe because it wants to try a reconnect, try to connect the websocket now.
+        EVLOG_debug << "Init and connect websocket for pending network slot: " << this->pending_network_slot;
+        if (this->websocket != nullptr) {
+            this->websocket->reconnect();
+        } else {
+            if (this->init_websocket(this->pending_network_slot) && this->websocket != nullptr) {
+                this->websocket->connect();
+            } else {
+                EVLOG_warning << "Could not initialize websocket with pending network slot: "
+                              << this->pending_network_slot;
+                // TODO what to do here???
+            }
+        }
+    }
+
+    config_slot_lock.unlock();
+
+    // Wait until a reconnect is requested for whatever reason (normally someone called
+    // 'reconnect_condition_variable.notify_<something>').
+    std::unique_lock<std::mutex> lock(reconnect_mutex);
+    // Return when 'running' is set to false (we want to stop the thread) or 'try_reconnect' is true (we want to
+    // reconnect to the websocket).
+    // TODO is there a situation where we wait here forever??
+    reconnect_condition_variable.wait(lock, [this] { return !this->running.load() || this->try_reconnect.load(); });
+    // At this point we want to re-run this function. Since 'run' is called in a while loop, it will be called again
+    // directly after returning this function.
+}
+
+bool ConnectivityManager::init_websocket(std::optional<int32_t> config_slot) {
+    std::string configuration_slot;
+    if (this->device_model.get_value<std::string>(ControllerComponentVariables::ChargePointId).find(':') !=
+        std::string::npos) {
+        EVLOG_AND_THROW(std::runtime_error("ChargePointId must not contain \':\'"));
+    }
+
+    int32_t config_slot_int;
+    if (config_slot.has_value()) {
+        EVLOG_debug << "Init websocket using configuration slot " << config_slot.value();
+        configuration_slot = std::to_string(config_slot.value());
+        config_slot_int = config_slot.value();
+    } else {
+        // No config slot given. Get first available config slot.
+        const std::vector<std::string> config_slot_vector = ocpp::get_vector_from_csv(
+            this->device_model.get_value<std::string>(ControllerComponentVariables::NetworkConfigurationPriority));
+        if (config_slot_vector.size() > 0) {
+            configuration_slot = config_slot_vector.at(0);
+            config_slot_int = std::stoi(configuration_slot);
+            EVLOG_debug << "Init websocket without config slot, use first available: " << config_slot_int;
+        } else {
+            // No network connection profile. Retry connecting after some time, maybe it is set manually.
+            EVLOG_info << "No network connection profile set. Try again later.";
```

Comment:
Maybe we can try this more often then every 300 seconds but not print this message every time?